qiskit/transpiler/passes/optimization/optimize_cliffords.py

```diff
@@ -0,0 +1,87 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Combine consecutive Cliffords over the same qubits."""
+
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.quantum_info.operators import Clifford
+
+
+class OptimizeCliffords(TransformationPass):
+    """Combine consecutive Cliffords over the same qubits.
+    This serves as an example of extra capabilities enabled by storing
+    Cliffords natively on the circuit.
+    """
+
+    def run(self, dag):
+        """Run the OptimizeCliffords pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
+        """
+
+        blocks = []
+        prev_node = None
+        cur_block = []
+
+        # Iterate over all nodes and collect consecutive Cliffords over the
+        # same qubits. In this very first proof-of-concept implementation
+        # we require the same ordering of qubits, but this restriction will
+        # be shortly removed. An interesting question is whether we may also
+        # want to compose Cliffords over different sets of qubits, such as
+        # cliff1 over qubits [1, 2, 3] and cliff2 over [2, 3, 4].
+        for node in dag.op_nodes():
```

Comment:
Is this algorithm sensitive to iteration order? Using `op_nodes()` this will return a list of op nodes from the dag in `node_id` order (which is nominally insertion order except for when there are intermediate removals). This could lead to getting a weird iteration order in some cases depending on how the dag was constructed. Do you want instead to iterate over in topological order to ensure we're looking at neighboring nodes on a qubit for building the block?

```suggestion
        for node in dag.topological_op_nodes():
```

Reply:
This is a good suggestion, thanks! Committed this change and a small test to see this in [0d583ed](https://github.com/Qiskit/qiskit-terra/pull/7966/commits/0d583edceab35c35e63c6b1d29aa81e0f742f3f5).